server/routerlicious/packages/lambdas/src/broadcaster/lambda.ts

```diff
@@ -113,23 +113,32 @@ export class BroadcasterLambda implements IPartitionLambda {
         }
 
         // Invoke the next send after a delay to give IO time to create more batches
-        this.messageSendingTimerId = taskScheduleFunction(() => {
+        this.messageSendingTimerId = taskScheduleFunction(async () => {
             const batchOffset = this.pendingOffset;
 
             this.current = this.pending;
             this.pending = new Map<string, BroadcasterBatch>();
             this.pendingOffset = undefined;
 
-            this.messageSendingTimerId = undefined;
-
             // Process all the batches + checkpoint
-            this.current.forEach((batch, topic) => {
+            const promises: Promise<void>[] = [];
+
+            for (const [topic, batch] of this.current) {
                 if (this.publisher.emit) {
-                    this.publisher.emit(topic, batch.event, batch.documentId, batch.messages);
+                    promises.push(this.publisher.emit(topic, batch.event, batch.documentId, batch.messages));
                 } else {
                     this.publisher.to(topic).emit(batch.event, batch.documentId, batch.messages);
                 }
-            });
+            }
+
+            try {
+                await Promise.all(promises);
+            } catch (ex) {
+                this.context.error(ex, { restart: true });
+                return;
+            }
```

Comment:
Should we refactor it this way?

if (this.publisher.emit) {
  for (const [topic, batch] of this.current) {
    // Do all promise stuff
  }
} else {
    for (const [topic, batch] of this.current) {
      // do regular stuff
  }
}

Reply:
Yea that sounds good